components/movementsensor/imuwit/imu.go

```diff
@@ -124,26 +124,51 @@ func (imu *wit) getMagnetometer() (r3.Vector, error) {
 }
 
 func (imu *wit) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	var err error
+
 	imu.mu.Lock()
 	defer imu.mu.Unlock()
 
-	var err error
-	// this only works when the imu is level to the surface of the earth, no inclines
+	// this will compensate for a tilted IMU if the tilt is less than 45 degrees
 	// do not let the imu near permanent magnets or things that make a strong magnetic field
-	imu.compassheading = calculateCompassHeading(imu.magnetometer.X, imu.magnetometer.Y)
+	imu.compassheading = imu.calculateCompassHeading()
 
 	return imu.compassheading, err
 }
 
-func calculateCompassHeading(x, y float64) float64 {
+// Helper function to calculate compass heading with tilt compensation.
+func (imu *wit) calculateCompassHeading() float64 {
+	pitch := imu.orientation.Pitch
+	roll := imu.orientation.Roll
+
+	var x, y float64
+
+	// Tilt compensation only works if the pitch and roll are between -45 and 45 degrees.
+	if roll >= -0.785 && roll <= 0.785 && pitch >= -0.785 && pitch <= 0.785 {
```

Comment:
you can do `math.Abs(roll)` and `math.Abs(pitch)` less than 45deg if you'd like for fewer statements.  Maybe also make this a constant so t's read and not re-culcated (fortyfivedegsInRad to something better)